specs/core/1_shard-data-chains.md

```diff
@@ -157,57 +161,73 @@ def get_switchover_epoch(state: BeaconState, epoch: Epoch, index: ValidatorIndex
+    def active_and_balance_filter(i: ValidatorIndex) -> bool:
+        if not is_active_validator(state.validators[i], epoch):
+            return False 
+        active_threshold = MAX_EFFECTIVE_BALANCE * seed[i % 32] // MAX_RANDOM_BYTE
+        if state.validators[i].effective_balance < active_threshold:
+            return False
+        return True
 
     # Take not-yet-cycled-out validators from earlier committee and already-cycled-in validators from
     # later committee; return a sorted list of the union of the two, deduplicated
-    return sorted(list(set(
-        [i for i in earlier_committee if epoch % PERSISTENT_COMMITTEE_PERIOD < get_switchover_epoch(state, epoch, i)]
-        + [i for i in later_committee if epoch % PERSISTENT_COMMITTEE_PERIOD >= get_switchover_epoch(state, epoch, i)]
-    )))
+    combined_committee = sorted(set(
+        [i for i in earlier_committee if before_switch(i) and active_and_balance_filter(i)]
+        + [i for i in later_committee if (not before_switch(i)) and active_and_balance_filter(i)]
+    ))
+
+    committee_size = min(len(combined_committee), SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH)
+    return [
+        compute_shuffled_index(combined_committee[i], len(combined_committee), seed)
+        for i in range(committee_size)
+    ]
 ```
 
-### `get_shard_proposer_index`
+### `get_shard_block_proposer_index`
 
 ```python
-def get_shard_proposer_index(state: BeaconState,
-                             shard: Shard,
-                             slot: Slot) -> Optional[ValidatorIndex]:
-    # Randomly shift persistent committee
-    persistent_committee = list(get_persistent_committee(state, shard, slot))
-    seed = hash(state.current_shuffling_seed + int_to_bytes(shard, length=8) + int_to_bytes(slot, length=8))
-    random_index = bytes_to_int(seed[0:8]) % len(persistent_committee)
-    persistent_committee = persistent_committee[random_index:] + persistent_committee[:random_index]
-
-    # Search for an active proposer
-    for index in persistent_committee:
-        if is_active_validator(state.validators[index], get_current_epoch(state)):
-            return index
-
-    # No block can be proposed if no validator is active
-    return None
+def get_shard_block_proposer_index(state: BeaconState,
+                                   shard: Shard,
+                                   slot: ShardSlot) -> Optional[ValidatorIndex]:
+    epoch_committee = get_shard_epoch_committee(state, shard, compute_epoch_of_shard_slot(slot))
+    if len(epoch_committee) == 0:
+        return None
+    else:
+        return epoch_committee[slot % len(epoch_committee)]
+```
+
+### `get_shard_block_attester_committee`
+
+```python
+def get_shard_block_attester_committee(state: BeaconState,
+                                       shard: Shard,
+                                       slot: ShardSlot) -> Sequence[Optional[ValidatorIndex]]:
+    committee_size = min(
```

Comment:
Comment saying the previous `committee_size` proposers attest to `slot` (inclusive of current slot) would be valuable